File: ModSources/InfiniCrafterLocal/Common/VFX/RuntimeColorPolicy.cs
#nullable enable
using Microsoft.Xna.Framework;
using System.IO;

namespace InfiniCrafterLocal.Common.VFX;

/// <summary>
/// Exact bounded color vocabulary shared by generated item lights and VFX.
/// Rich visual prose/palettes never enter this runtime selector.
/// </summary>
public static class RuntimeColorPolicy
{
    public static string NormalizeRequired(string? value, bool allowEmpty = false)
    {
        string token = (value ?? "").Trim().ToLowerInvariant();
        if (allowEmpty && token.Length == 0)
            return token;
        if (token is "white" or "red" or "orange" or "yellow" or "green" or "cyan" or
            "blue" or "purple" or "pink" or "gray" or "black")
            return token;
        throw new InvalidDataException($"unknown runtime color '{token}'");
    }

    private static string NormalizeForRendering(string? value)
    {
        string token = (value ?? "").Trim().ToLowerInvariant();
        return token is "white" or "gray" or "brown" or "tan" or "red" or "orange" or
            "yellow" or "gold" or "green" or "cyan" or "blue" or "purple" or "pink" or "black"
            ? token
            : "";
    }

    public static Color Resolve(string? value, Color fallback)
        => NormalizeForRendering(value) switch
        {
            "white" => new Color(235, 235, 235),
            "gray" => new Color(170, 170, 180),
            "brown" => new Color(150, 100, 65),
            "tan" => new Color(225, 190, 120),
            "red" => new Color(255, 85, 85),
            "orange" => new Color(255, 155, 70),
            "yellow" => new Color(255, 235, 90),
            "gold" => new Color(255, 205, 70),
            "green" => new Color(110, 255, 145),
            "cyan" => new Color(90, 235, 255),
            "blue" => new Color(120, 190, 255),
            "purple" => new Color(190, 110, 255),
            "pink" => new Color(255, 145, 215),
            "black" => new Color(25, 25, 30),
            _ => fallback,
        };
}

File: LocalGenerator/infini_local/core/runtime_color_policy.py
from __future__ import annotations

from typing import Any

# Exact current runtime/presentation color vocabulary.  Rich image palettes stay in
# visual prompts; these names are only for bounded C# light/dust/tint execution.
CANONICAL_RUNTIME_COLORS = frozenset({
    "white", "gray", "brown", "tan", "red", "orange", "yellow",
    "gold", "green", "cyan", "blue", "purple", "pink",
})

_EFFECT_RUNTIME_COLOR = {
    "none": "white",
    "electric": "cyan",
    "slime": "green",
    "star": "gold",
    "flame": "orange",
    "frost": "blue",
    "leaf": "green",
    "shadow": "purple",
    "poison": "green",
    "blood": "red",
    "honey": "gold",
    "sand": "tan",
    "lunar": "cyan",
    "heal": "pink",
    "holy": "gold",
    "smoke": "gray",
}


def normalize_runtime_color(value: Any, fallback: str = "") -> str:
    token = str(value or "").strip().lower()
    return token if token in CANONICAL_RUNTIME_COLORS else fallback


def runtime_color_for_effect(effect: Any) -> str:
    return _EFFECT_RUNTIME_COLOR.get(str(effect or "none").strip().lower(), "white")


__all__ = ["CANONICAL_RUNTIME_COLORS", "normalize_runtime_color", "runtime_color_for_effect"]
